README.md
# Kaolin-inspiried themes for VS Code/Codium

[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-green.svg)](http://www.gnu.org/licenses/gpl-3.0)

![Preview](https://raw.githubusercontent.com/zed-nait/kaolin-vscode-themes/master/images/preview.png)

## Installation

1. Open **Extensions** sidebar panel in VS Code. `View → Extensions`
2. Search for `kaolin themes`
3. Click **Install** to install it
4. Code > Preferences > Color Theme > **Kaolin**

## Contribute

1. Clone and open this [repo](https://github.com/zed-nait/kaolin-vscode-themes) in VS Code
2. Run `yarn` to install the dependencies.
3. Press `F5` to open a new window with your extension loaded
4. Open `Code > Preferences > Color Theme` **"Kaolin ..."**
5. Make changes to the [`/src/theme.js`](https://github.com/zed-nait/kaolin-vscode-themes/blob/master/src/theme.js) file.
   - **UI**: For all changes to the "outer UI", like (status bar, file navigation etc.), take a look at the [Theme Color](https://code.visualstudio.com/api/references/theme-color) reference.
   - **Syntax**: For changes to the "code highlighting", examine the syntax scopes by invoking the [`Developer: Inspect Editor Tokens and Scopes`](https://code.visualstudio.com/api/language-extensions/syntax-highlight-guide#scope-inspector) command from the Command Palette in the Extension Development Host window.

If you are making a Pull Request, Please give me a screenshot of before/after!

## Credits

- [@ogdenwebb](https://github.com/ogdenwebb) - awesome GNU Emacs [color theme](https://github.com/ogdenwebb/emacs-kaolin-themes), that made this project possible
- [@katran](https://t.me/katran_tg) - beautiful logo


CHANGELOG.md
# kaolin-vscode-themes

## 0.1.0

- Initial release


src/theme.js
const { getColors } = require("./colors");

function getTheme({ theme, name, type }) {
  const {
    bg0,
    bg1,
    bg2,
    bg3,
    bg4,
    fg1,
    fg2,
    fg3,
    fg4,
    link,
    error,
    button,
    checkbox,
    dropdown,
    input,
    debug,
    syntax,
    activeBar,
    sideBar,
    selection,
    search,
    badge,
    progressBar,
    ansi,
    indent,
    warning,
    done,
    pickerGroup,
    statusBar,
    tab,
    highlight,
    cursor,
    diff,
    bracketHighlight,
    peekView,
    merge,
  } = getColors(theme);

  const colors = {
    focusBorder: bg4,
    foreground: fg1,
    descriptionForeground: fg1,
    errorForeground: error,

    "textLink.foreground": link,
    "textLink.activeForeground": link,
    "textBlockQuote.background": bg0,
    "textBlockQuote.border": bg1,
    "textCodeBlock.background": bg2,
    "textSeparator.foreground": bg3,

    "icon.foreground": fg2,

    "keybindingLabel.foreground": fg2,

    "button.background": button.bg,
    "button.foreground": button.fg,
    "button.hoverBackground": "",

    "button.secondaryBackground": button.bg,
    "button.secondaryForeground": button.fg,
    "button.secondaryHoverBackground": "",

    "checkbox.background": checkbox.bg,
    "checkbox.foreground": checkbox.fg,
    "checkbox.border": checkbox.border,

    "dropdown.background": dropdown.bg,
    "dropdown.border": dropdown.border,
    "dropdown.foreground": dropdown.fg,
    "dropdown.listBackground": dropdown.listBackground,

    "input.background": input.bg,
    "input.border": input.border,
    "input.foreground": input.fg,
    "input.placeholderForeground": input.placeholderFg,

    "badge.foreground": badge.fg,
    "badge.background": badge.bg,

    "progressBar.background": progressBar,

    "titleBar.activeForeground": "",
    "titleBar.activeBackground": "",
    "titleBar.inactiveForeground": "",
    "titleBar.inactiveBackground": "",
    "titleBar.border": "",

    "activityBar.activeBackground": activeBar.activeBg,
    "activityBar.activeBorder": activeBar.activeBorder,
    "activityBar.background": activeBar.bg,
    "activityBar.border": activeBar.border,
    "activityBar.foreground": activeBar.fg,
    "activityBar.inactiveForeground": fg4,

    "activityBarBadge.foreground": badge.fg,
    "activityBarBadge.background": badge.bg,

    "sideBar.foreground": sideBar.fg,
    "sideBar.background": sideBar.bg,
    "sideBar.border": sideBar.border,
    "sideBarTitle.foreground": fg2,
    "sideBarSectionHeader.foreground": fg1,
    "sideBarSectionHeader.background": bg3,
    "sideBarSectionHeader.border": bg4,

    "list.activeSelectionBackground": selection,
    "list.activeSelectionForeground": fg2,
    "list.focusBackground": bg3,
    "list.focusForeground": fg4,
    "list.highlightForeground": error,
    "list.hoverBackground": bg3,
    "list.hoverForeground": fg2,
    "list.inactiveFocusBackground": bg3,
    "list.inactiveSelectionBackground": bg3,
    "list.inactiveSelectionForeground": fg2,

    "tree.indentGuidesStroke": indent,

    "notificationCenterHeader.foreground": fg2,
    "notificationCenterHeader.background": bg2,
    "notifications.foreground": fg3,
    "notifications.background": bg1,
    "notifications.border": bg3,
    "notificationsErrorIcon.foreground": error,
    "notificationsWarningIcon.foreground": warning,
    "notificationsInfoIcon.foreground": done,

    "pickerGroup.border": pickerGroup.border,
    "pickerGroup.foreground": pickerGroup.fg,
    "quickInput.background": bg1,
    "quickInput.foreground": fg2,

    "statusBar.foreground": statusBar.default.fg,
    "statusBar.background": statusBar.default.bg,
    "statusBar.border": statusBar.default.border,
    "statusBar.noFolderBackground": statusBar.nofolder.bg,
    "statusBar.noFolderForeground": statusBar.nofolder.fg,
    "statusBar.debuggingForeground": statusBar.debug.fg,
    "statusBar.debuggingBackground": statusBar.debug.bg,
    "statusBarItem.remoteForeground": statusBar.remote.fg,
    "statusBarItem.remoteBackground": statusBar.remote.bg,

    "editorGroupHeader.tabsBackground": bg0,
    "editorGroupHeader.tabsBorder": bg3,
    "editorGroup.border": bg3,

    "tab.activeForeground": tab.active.fg,
    "tab.activeBackground": tab.active.bg,
    "tab.activeBorder": tab.active.border,
    "tab.activeBorderTop": tab.active.borderTop,
    "tab.border": tab.border,
    "tab.inactiveForeground": tab.default.fg,
    "tab.inactiveBackground": tab.default.bg,
    "tab.hoverBackground": tab.hover.bg,
    "tab.hoverForeground": tab.hover.fg,
    "tab.unfocusedActiveBorder": "",
    "tab.unfocusedHoverBackground": "",
    "tab.unfocusedActiveBorderTop": "",

    "breadcrumb.foreground": fg4,
    "breadcrumb.focusForeground": fg1,
    "breadcrumb.activeSelectionForeground": fg1,
    "breadcrumbPicker.background": bg1,

    "editorWidget.background": bg1,

    "editor.foreground": fg1,
    "editor.background": bg1,
    "editor.foldBackground": `${bg2}80`,
    "editor.lineHighlightBackground": bg2,
    // "editor.lineHighlightBorder": "",
    "editorLineNumber.foreground": syntax.comment,
    "editorLineNumber.activeForeground": highlight,
    "editorIndentGuide.background": `${indent}65`,
    "editorIndentGuide.activeBackground": indent,
    "editorWhitespace.foreground": indent,
    "editorCursor.foreground": cursor,
    "editorLink.activeForeground": link,

    "editor.findMatchBackground": `${search}40`,
    "editor.findMatchHighlightBackground": `${search}30`,
    "editor.rangeHighlightBackground": `${search}30`,
    "editor.linkedEditingBackground": "",
    "editor.inactiveSelectionBackground": `${selection}90`,
    "editor.selectionBackground": selection,
    "editor.selectionHighlightBackground": selection,

    "editorBracketMatch.background": selection,
    "editorBracketMatch.border": fg4,

    "editorGutter.modifiedBackground": diff.modified,
    "editorGutter.addedBackground": diff.added,
    "editorGutter.deletedBackground": diff.deleted,

    "diffEditor.insertedLineBackground": `${diff.added}25`,
    "diffEditor.removedTextBackground": `${diff.deleted}25`,

    "scrollbarSlider.background": `${fg4}70`,
    "scrollbarSlider.hoverBackground": `${fg4}90`,
    "scrollbarSlider.activeBackground": fg4,

    "editorOverviewRuler.border": ansi.red,

    // "panel.background": "",
    // "panel.border": "",
    // "panelTitle.activeBorder": "",
    // "panelTitle.activeForeground": "",
    // "panelTitle.inactiveForeground": "",
    // "panelInput.border": "",

    "debugIcon.breakpointForeground": debug.breakpoint,
    "debugConsole.infoForeground": debug.info,
    "debugConsole.warningForeground": debug.warning,
    "debugConsole.errorForeground": debug.error,
    "debugConsole.sourceForeground": debug.source,
    "debugConsoleInputIcon.foreground": debug.inputIcon,
    "debugToolBar.background": debug.toolBar,

    "debugTokenExpression.name": syntax.class,
    "debugTokenExpression.value": syntax.variable,
    "debugTokenExpression.string": syntax.string,
    "debugTokenExpression.boolean": syntax.bool,
    "debugTokenExpression.number": syntax.num,
    "debugTokenExpression.error": syntax.error,

    "terminal.foreground": fg4,
    "terminal.ansiBlack": ansi.black,
    "terminal.ansiRed": ansi.red,
    "terminal.ansiGreen": ansi.green,
    "terminal.ansiYellow": ansi.yellow,
    "terminal.ansiBlue": ansi.blue,
    "terminal.ansiMagenta": ansi.magenta,
    "terminal.ansiCyan": ansi.cyan,
    "terminal.ansiWhite": ansi.white,
    "terminal.ansiBrightBlack": ansi.brightBlack,
    "terminal.ansiBrightRed": ansi.brightRed,
    "terminal.ansiBrightGreen": ansi.brightGreen,
    "terminal.ansiBrightYellow": ansi.brightYellow,
    "terminal.ansiBrightBlue": ansi.brightBlue,
    "terminal.ansiBrightMagenta": ansi.brightMagenta,
    "terminal.ansiBrightCyan": ansi.brightCyan,
    "terminal.ansiBrightWhite": ansi.brightWhite,

    "editorBracketHighlight.foreground1": bracketHighlight[0],
    "editorBracketHighlight.foreground2": bracketHighlight[1],
    "editorBracketHighlight.foreground3": bracketHighlight[2],
    "editorBracketHighlight.foreground4": bracketHighlight[3],
    "editorBracketHighlight.foreground5": bracketHighlight[4],
    "editorBracketHighlight.foreground6": bracketHighlight[5],
    "editorBracketHighlight.unexpectedBracket.foreground": error,

    "gitDecoration.addedResourceForeground": diff.added,
    "gitDecoration.modifiedResourceForeground": diff.modified,
    "gitDecoration.deletedResourceForeground": diff.deleted,
    "gitDecoration.untrackedResourceForeground": diff.untracked,
    "gitDecoration.ignoredResourceForeground": diff.ignored,
    "gitDecoration.conflictingResourceForeground": diff.conflicting,
    "gitDecoration.submoduleResourceForeground": diff.submodule,

    "peekViewEditor.background": peekView.editorBackground,
    "peekViewEditor.matchHighlightBackground": peekView.editorMatchHighlight,
    "peekViewResult.background": peekView.resultBackground,
    "peekViewResult.matchHighlightBackground": peekView.resultMatchHighlight,

    "merge.currentHeaderBackground": merge.currentHeader,
    "merge.currentContentBackground": merge.currentContent,
    "merge.incomingHeaderBackground": merge.incomingHeader,
    "merge.incomingContentBackground": merge.incomingContent,
    "merge.commonContentBackground": merge.commonContent,
    "merge.commonHeaderBackground": merge.commonHeader,
  };

  const tokenColors = [
    {
      scope: ["keyword"],
      settings: {
        foreground: syntax.keyword,
      },
    },
    {
      scope: ["entity.name.class", "variable.other.enummember"],
      settings: {
        foreground: syntax.class,
      },
    },
    {
      scope: ["storage.type"],
      settings: {
        foreground: syntax.keyword,
      },
    },
    {
      scope: ["variable"],
      settings: {
        foreground: syntax.variable,
      },
    },
    {
      scope: ["string"],
      settings: {
        foreground: syntax.string,
      },
    },
    {
      scope: ["punctuation.section.braces.begin"],
      settings: {
        foreground: syntax.class,
      },
    },
    {
      scope: ["support.type"],
      settings: {
        foreground: syntax.type,
      },
    },
    {
      scope: ["comment.line", "comment.block"],
      settings: {
        foreground: syntax.comment,
      },
    },
    {
      scope: ["punctuation.decorator", "entity.name.function"],
      settings: {
        foreground: syntax.num,
      },
    },
    {
      scope: ["constant.language.boolean"],
      settings: {
        foreground: syntax.bool,
      },
    },
    {
      scope: ["variable.other.readwrite"],
      settings: {
        foreground: fg1,
      },
    },
    {
      scope: ["meta.diff.header.git", "meta.diff.index.git"],
      settings: {
        foreground: diff.modified,
      },
    },
    {
      scope: ["meta.diff.header.from-file", "markup.deleted.diff"],
      settings: {
        foreground: diff.deleted,
      },
    },
    {
      scope: ["meta.diff.header.to-file", "markup.inserted.diff"],
      settings: {
        foreground: diff.added,
      },
    },
    {
      scope: ["meta.diff.range.unified"],
      settings: {
        foreground: diff.range,
      },
    },
    {
      scope: ["source.diff"],
      settings: {
        foreground: diff.fg,
      },
    },
    {
      scope: ["constant.numeric"],
      settings: {
        foreground: syntax.num,
      },
    },
    {
      scope: ["entity.name.tag"],
      settings: {
        foreground: syntax.variable,
      },
    },
    {
      scope: ["meta.attribute.style"],
      settings: {
        foreground: syntax.type,
      },
    },
  ];

  return {
    name: name,
    type,
    semanticHighlighting: true,
    colors: Object.entries(colors).reduce(
      (acc, [key, value]) => (value !== "" ? { ...acc, [key]: value } : acc),
      {}
    ),
    tokenColors,
  };
}

module.exports = getTheme;


src/showcase.js
const fs = require("fs").promises;

const lib = require("../colors/kaolin-themes-lib.json");
const light = require("../colors/kaolin-themes-light.json");
const dark = require("../colors/kaolin-themes-dark.json");

function generate(colors) {
  return `<table style="width:20rem;table-layout:auto">
  <thead>
    <tr>
      <th> Name </th>
      <th> HEX </th>
      <th> Colored Background </th>
    </tr>
  </thead>
  <tbody>${Object.entries(colors)
    .map(
      ([key, value]) => `
      <tr>
        <td> ${key} </td>
        <td> ${value} </td>
        <td style="background-color:${value}" />
      </tr>`
    )
    .join("")}</tbody>
</table>`;
}

fs.mkdir("./showcase", { recursive: true }).then(() =>
  Promise.all([
    fs.writeFile("./showcase/colors-light.md", generate({ ...lib, ...light })),
    fs.writeFile("./showcase/colors-dark.md", generate({ ...lib, ...dark })),
  ])
);


src/colors.js
const lib = require("../colors/kaolin-themes-lib.json");
const light = require("../colors/kaolin-themes-light.json");

const lightColor = (altBg) => {
  let c = { ...lib, ...light };

  const fg0 = c.white0;
  const fg1 = c.gray1;
  const fg2 = c.gray2;
  const fg3 = c.gray3;
  const fg4 = c.gray4;

  const bg0 = "#f5f6f5";
  const bg1 = altBg ? "#fbfbfb" : "#edeeeb";
  const bg2 = altBg ? c.white0 : "#dfe1dc";
  const bg3 = altBg ? c.white1 : "#d1d4cd";
  const bg4 = altBg ? c.white2 : "#c8ccc3";

  const info = c.azure4;
  const error = c.red3;
  const warning = c.orange1;
  const selection = c.azure9;

  return {
    fg0,
    fg1,
    fg2,
    fg3,
    fg4,
    bg0,
    bg1,
    bg2,
    bg3,
    bg4,
    selection,
    info,
    error,
    warning,
    highlight: c.green2,
    indent: c.gray9,
    done: c.teal2,
    link: c.ultramarine1,
    progressBar: c.cyan3,
    cursor: c.gray3,
    search: c.teal3,
    button: {
      fg: bg1,
      bg: c.teal2,
    },
    badge: {
      fg: fg0,
      bg: c.red3,
    },
    indent: c.gray9,
    debug: {
      fg: fg0,
      bg: c.red3,
    },
    diff: {
      added: c.aquamarine2,
      modified: c.purple2,
      deleted: c.red1,
      untracked: c["spring-green2"],
      ignored: c.gray8,
      conflicting: error,
      submodule: c.azure1,
      range: c.amber2,
      fg: fg1,
    },
    syntax: {
      comment: c.gray7,
      constant: c.magenta4,
      keyword: c.teal2,
      string: c.erin2,
      variable: c.magenta4,
      class: c.vermilion3,
      type: c.azure4,
      num: c.red1,
      bool: c.red1,
      error,
    },
    ansi: {
      black: fg1,
      brightBlack: fg3,
      red: c.red3,
      brightRed: c.red3,
      green: c.teal2,
      brightGreen: c["spring-green2"],
      yellow: c.vermilion3,
      brightYellow: c.orange1,
      blue: c.azure1,
      brightBlue: c.azure4,
      magenta: c.cerise4,
      brightMagenta: c.ultramarine1,
      cyan: c.cyan3,
      brightCyan: c.cyan2,
      white: bg4,
      brightWhite: bg0,
    },
    tab: {
      border: bg4,
      default: {
        bg: bg0,
        fg: fg3,
      },
      active: {
        bg: bg1,
        fg: fg3,
        border: bg2,
        borderTop: bg4,
      },
      hover: {
        bg: bg2,
        fg: fg2,
      },
    },
    checkbox: {
      bg: bg0,
      fg: c.azure4,
      border: bg4,
    },
    dropdown: {
      bg: bg0,
      fg: fg4,
      border: bg4,
      listBackground: bg0,
    },
    input: {
      bg: bg0,
      border: bg4,
      fg: fg1,
      placeholderFg: `${fg1}90`,
    },
    statusBar: {
      default: {
        bg: c["spring-green4"],
        fg: bg2,
        border: bg3,
      },
      debug: {
        bg: c.crimson4,
        fg: bg1,
      },
      remote: {
        bg: c.azure4,
        fg: bg1,
      },
      nofolder: {
        bg: c.ultramarine4,
        fg: bg1,
      },
    },
    activeBar: {
      bg: bg2,
      fg: fg2,
      border: bg4,
      activeBg: bg4,
      activeBorder: bg4,
    },
    sideBar: {
      fg: fg2,
      bg: bg2,
      border: bg3,
    },
    pickerGroup: {
      fg: c.violet2,
      border: bg2,
    },
    bracketHighlight: [c.teal1, c.cerise4, c.azure4, c.ultramarine4, c.crimson4, c.vermilion4],
    debug: {
      breakpoint: c.red3,
      info: info,
      warning: warning,
      error: error,
      source: fg2,
      inputIcon: c.cerise4,
      toolBar: bg0,
    },
    peekView: {
      editorBackground: bg1,
      editorMatchHighlight: `${selection}80`,
      resultBackground: bg2,
      resultMatchHighlight: `${c.amber0}80`,
    },
    merge: {
      currentHeader: `${c.aquamarine4}60`,
      currentContent: `${c.aquamarine4}25`,
      incomingHeader: `${c.azure1}60`,
      incomingContent: `${c.azure1}25`,
      commonHeader: bg3,
      commonContent: bg2,
    },
  };
};

function getColors(theme) {
  switch (theme) {
    case "light":
      return lightColor(false);
    case "light-alt":
      return lightColor(true);
    default:
      throw new Error(`Colors are missing for value: ${theme}`);
  }
}

module.exports = { getColors };
